# Data Wrangling, Analysis and Report on @dog_rates twitter feeds

## Content

#### 1. Gather data
   >> `Twitter Archives`
    
   >> `Image Predictions`
    
   >> `Tweets json`

#### 2.  Assess data
>>`Summary of Assessments:`
>>>      Quality
>>>      Tidiness
    

#### 3. Clean data

>> `Define`

>> `Code`

>> `Test`

#### 4. Analyse data
     
   >> `Feature Enginnering`
     
  >>`Visualization`

#### 5.  Report
   >> `wrangle_report`
   
   >>`act_report`

> 

>

### Gathering Data

> Import Required Libraries and Modules

In [333]:
import pandas as pd
import requests
import json
import os

> Dataset 1:  Twitter Archives

In [334]:
tweet_archive_enhanced = './datasets/twitter-archive-enhanced.csv'

In [335]:
df_twitter_archive = pd.read_csv(tweet_archive_enhanced)
df_twitter_archive.head(1)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   

   rating_denominator     name doggo floofer pupper puppo  
0                  10  Phineas  None    None   None  None

In [336]:
#size of tweets where rating was included
df_twitter_archive.nunique()[0]

2356

> Dataset 2: Image Predictions

In [337]:
img_predictions_link = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'

In [338]:
res = requests.get(img_predictions_link)

In [339]:
#walk through datasets directory
!cd datasets
!ls

analysis  README.md  requirements.txt  twitter-api.ipynb
datasets  report     resources	       twitter-api.py


In [340]:
#create a new file to save dowmloaded data
!touch img_pred.tsv
!mv img_pred.tsv ~ datasets

mv: cannot move '/home/patrick' to a subdirectory of itself, 'datasets/patrick'


In [341]:
# write downloaded data into file
with open ('datasets/img_pred.tsv', mode='wb') as file:
    file.write(res.content)

In [342]:
df_img_pred = pd.read_csv('datasets/img_pred.tsv', sep='\t')
df_img_pred.head(2)

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True

> Dataset 3: Json Tweets

In [343]:
tweet_json = 'datasets/tweet-json-2.txt'

In [344]:
tweets =[]
with open (tweet_json, mode='rb') as file:
        for line in file:
            tweet = json.loads(line) 
            tweets.append(tweet)

In [345]:
df_tweets = pd.DataFrame(tweets)

> read this .txt file line by line into a pandas DataFrame with (at minimum) tweet ID, retweet count, and favorite count.

In [346]:
df_tweets.head(1)

created_at                  id              id_str  \
0  Tue Aug 01 16:23:56 +0000 2017  892420643555336193  892420643555336193   

                                           full_text  truncated  \
0  This is Phineas. He's a mystical boy. Only eve...      False   

  display_text_range                                           entities  \
0            [0, 85]  {'hashtags': [], 'symbols': [], 'user_mentions...   

                                   extended_entities  \
0  {'media': [{'id': 892420639486877696, 'id_str'...   

                                              source  in_reply_to_status_id  \
0  <a href="http://twitter.com/download/iphone" r...                    NaN   

   ... favorite_count  favorited retweeted possibly_sensitive  \
0  ...          39467      False     False              False   

  possibly_sensitive_appealable lang retweeted_status quoted_status_id  \
0                         False   en              NaN              NaN   

  quoted_status_id_str  quoted_status  
0                  NaN            NaN  

[1 rows x 31 columns]

### 

### Assessing Data

##### After gathering each of the above pieces of data, assess them visually and programmatically for quality and tidiness issues. 
##### Detect and document at least eight `(8) quality issues and two (2) tidiness issues` in your wrangle_act.ipynb Jupyter Notebook. 
##### To meet specifications, the issues that satisfy the Project Motivation (see the Key Points header on the previous page) must be assessed.

> Aseesment 1:  Twitter Archive

In [347]:
df_twitter_archive

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
...                  ...                    ...                  ...   
2351  666049248165822465                    NaN                  NaN   
2352  666044226329800704                    NaN                  NaN   
2353  666033412701032449                    NaN                  NaN   
2354  666029285002620928                    NaN                  NaN   
2355  666020888022790149                    NaN                  NaN   

                      timestamp  \
0     2017-08-01 16:23:56 +0000   
1     2017-08-01 00:17:27 +0000   
2     2017-07-31 00:18:03 +0000   
3     2017-07-30 15:58:51 +0000   
4     2017-07-29 16:00:24 +0000   
...                         ...   
2351  2015-11-16 00:24:50 +0000   
2352  2015-11-16 00:04:52 +0000   
2353  2015-11-15 23:21:54 +0000   
2354  2015-11-15 23:05:30 +0000   
2355  2015-11-15 22:32:08 +0000   

                                                 source  \
0     <a href="http://twitter.com/download/iphone" r...   
1     <a href="http://twitter.com/download/iphone" r...   
2     <a href="http://twitter.com/download/iphone" r...   
3     <a href="http://twitter.com/download/iphone" r...   
4     <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
2351  <a href="http://twitter.com/download/iphone" r...   
2352  <a href="http://twitter.com/download/iphone" r...   
2353  <a href="http://twitter.com/download/iphone" r...   
2354  <a href="http://twitter.com/download/iphone" r...   
2355  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
0     This is Phineas. He's a mystical boy. Only eve...                  NaN   
1     This is Tilly. She's just checking pup on you....                  NaN   
2     This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3     This is Darla. She commenced a snooze mid meal...                  NaN   
4     This is Franklin. He would like you to stop ca...                  NaN   
...                                                 ...                  ...   
2351  Here we have a 1949 1st generation vulpix. Enj...                  NaN   
2352  This is a purebred Piers Morgan. Loves to Netf...                  NaN   
2353  Here is a very happy pup. Big fan of well-main...                  NaN   
2354  This is a western brown Mitsubishi terrier. Up...                  NaN   
2355  Here we have a Japanese Irish Setter. Lost eye...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
0                          NaN                        NaN   
1                          NaN                        NaN   
2                          NaN                        NaN   
3                          NaN                        NaN   
4                          NaN                        NaN   
...                        ...                        ...   
2351                       NaN                        NaN   
2352                       NaN                        NaN   
2353                       NaN                        NaN   
2354                       NaN                        NaN   
2355                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
0     https://twitter.com/dog_rates/status/892420643...                13   
1     https://twitter.com/dog_rates/status/892177421...                13   
2     https://twitter.com/dog_rates/status/891815181...                12   
3     https://twitt

In [348]:
#check data size
df_twitter_archive.tweet_id.nunique()

2356

In [349]:
#gather general info about data
df_twitter_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

In [350]:
#take a look at the list of unique names
df_twitter_archive.name.unique()

array(['Phineas', 'Tilly', 'Archie', 'Darla', 'Franklin', 'None', 'Jax',
       'Zoey', 'Cassie', 'Koda', 'Bruno', 'Ted', 'Stuart', 'Oliver',
       'Jim', 'Zeke', 'Ralphus', 'Canela', 'Gerald', 'Jeffrey', 'such',
       'Maya', 'Mingus', 'Derek', 'Roscoe', 'Waffles', 'Jimbo', 'Maisey',
       'Lilly', 'Earl', 'Lola', 'Kevin', 'Yogi', 'Noah', 'Bella',
       'Grizzwald', 'Rusty', 'Gus', 'Stanley', 'Alfy', 'Koko', 'Rey',
       'Gary', 'a', 'Elliot', 'Louis', 'Jesse', 'Romeo', 'Bailey',
       'Duddles', 'Jack', 'Emmy', 'Steven', 'Beau', 'Snoopy', 'Shadow',
       'Terrance', 'Aja', 'Penny', 'Dante', 'Nelly', 'Ginger', 'Benedict',
       'Venti', 'Goose', 'Nugget', 'Cash', 'Coco', 'Jed', 'Sebastian',
       'Walter', 'Sierra', 'Monkey', 'Harry', 'Kody', 'Lassie', 'Rover',
       'Napolean', 'Dawn', 'Boomer', 'Cody', 'Rumble', 'Clifford',
       'quite', 'Dewey', 'Scout', 'Gizmo', 'Cooper', 'Harold', 'Shikha',
       'Jamesy', 'Lili', 'Sammy', 'Meatball', 'Paisley', 'Albus',
       'Nept

In [351]:
# check the counts of names
unique_names = (df_twitter_archive.name.value_counts())
unique_names[:20]

None       745
a           55
Charlie     12
Lucy        11
Oliver      11
Cooper      11
Tucker      10
Penny       10
Lola        10
Bo           9
Winston      9
Sadie        8
the          8
Toby         7
Bailey       7
Daisy        7
Buddy        7
an           7
Rusty        6
Stanley      6
Name: name, dtype: int64

In [352]:
len(unique_names)

957

In [353]:
# THere are invalid names such as None and a

In [354]:
# inspect column names where values is None
df_twitter_archive.query('name=="None"')

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
5     891087950875897856                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
...                  ...                    ...                  ...   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2351  666049248165822465                    NaN                  NaN   
2355  666020888022790149                    NaN                  NaN   

                      timestamp  \
5     2017-07-29 00:08:17 +0000   
7     2017-07-28 00:22:40 +0000   
12    2017-07-25 01:55:32 +0000   
24    2017-07-18 16:08:03 +0000   
25    2017-07-18 00:07:08 +0000   
...                         ...   
2342  2015-11-16 02:38:37 +0000   
2343  2015-11-16 01:59:36 +0000   
2344  2015-11-16 01:52:02 +0000   
2351  2015-11-16 00:24:50 +0000   
2355  2015-11-15 22:32:08 +0000   

                                                 source  \
5     <a href="http://twitter.com/download/iphone" r...   
7     <a href="http://twitter.com/download/iphone" r...   
12    <a href="http://twitter.com/download/iphone" r...   
24    <a href="http://twitter.com/download/iphone" r...   
25    <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
2342  <a href="http://twitter.com/download/iphone" r...   
2343  <a href="http://twitter.com/download/iphone" r...   
2344  <a href="http://twitter.com/download/iphone" r...   
2351  <a href="http://twitter.com/download/iphone" r...   
2355  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
5     Here we have a majestic great white breaching ...                  NaN   
7     When you watch your owner call another dog a g...                  NaN   
12    Here's a puppo that seems to be on the fence a...                  NaN   
24    You may not have known you needed to see this ...                  NaN   
25    This... is a Jubilant Antarctic House Bear. We...                  NaN   
...                                                 ...                  ...   
2342  Here we have a well-established sunblockerspan...                  NaN   
2343  Let's hope this flight isn't Malaysian (lol). ...                  NaN   
2344  Here we have a northern speckled Rhododendron....                  NaN   
2351  Here we have a 1949 1st generation vulpix. Enj...                  NaN   
2355  Here we have a Japanese Irish Setter. Lost eye...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
5                          NaN                        NaN   
7                          NaN                        NaN   
12                         NaN                        NaN   
24                         NaN                        NaN   
25                         NaN                        NaN   
...                        ...                        ...   
2342                       NaN                        NaN   
2343                       NaN                        NaN   
2344                       NaN                        NaN   
2351                       NaN                        NaN   
2355                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
5     https://twitter.com/dog_rates/status/891087950...                13   
7     https://twitter.com/dog_rates/status/890729181...                13   
12    https://twitter.com/dog_rates/status/889665388...                13   
24    https://twitt

> inspect text where name is None

In [355]:
df_twitter_archive.query('name=="None"')['text'].values.tolist()[:20]

["Here we have a majestic great white breaching off South Africa's coast. Absolutely h*ckin breathtaking. 13/10 (IG: tucker_marlo) #BarkWeek https://t.co/kQ04fDDRmh",
 "When you watch your owner call another dog a good boy but then they turn back to you and say you're a great boy. 13/10 https://t.co/v0nONBcwxq",
 "Here's a puppo that seems to be on the fence about something haha no but seriously someone help her. 13/10 https://t.co/BxvuXk0UCm",
 'You may not have known you needed to see this today. 13/10 please enjoy (IG: emmylouroo) https://t.co/WZqNqygEyV',
 'This... is a Jubilant Antarctic House Bear. We only rate dogs. Please only send dogs. Thank you... 12/10 would suffocate in floof https://t.co/4Ad1jzJSdp',
 '@NonWhiteHat @MayhewMayhem omg hello tanner you are a scary good boy 12/10 would pet with extreme caution',
 'RT @Athletics: 12/10 #BATP https://t.co/WxwJmvjfxo',
 'I have a new hero and his name is Howard. 14/10 https://t.co/gzLHboL7Sk',
 'Here we have a corgi undercover a

> inspect text where name is not None

In [356]:
df_twitter_archive.query('name!="None"')['text'].values.tolist()[:20]

["This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU",
 "This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boops, the whole bit. 13/10 https://t.co/0Xxu71qeIV",
 'This is Archie. He is a rare Norwegian Pouncing Corgo. Lives in the tall grass. You never know when one may strike. 12/10 https://t.co/wUnZnhtVJB',
 'This is Darla. She commenced a snooze mid meal. 13/10 happens to the best of us https://t.co/tD36da7qLQ',
 'This is Franklin. He would like you to stop calling him "cute." He is a very fierce shark and should be respected as such. 12/10 #BarkWeek https://t.co/AtUZn91f7f',
 'Meet Jax. He enjoys ice cream so much he gets nervous around it. 13/10 help Jax enjoy more things by clicking below\n\nhttps://t.co/Zr4hWfAs1H https://t.co/tVJBRMnhxl',
 "This is Zoey. She doesn't want to be one of the scary sharks. Just wants to be a snuggly pettable boatpet. 13/10 #BarkWee

##### From the above queries, we can observe rows where names of dog are not available do not have text in them

#### Inspect ratings (numerator & denominator)

In [357]:
ratings = df_twitter_archive[['rating_numerator', 'rating_denominator']]

In [358]:
ratings.rating_numerator.value_counts()

12      558
11      464
10      461
13      351
9       158
8       102
7        55
14       54
5        37
6        32
3        19
4        17
1         9
2         9
420       2
0         2
15        2
75        2
80        1
20        1
24        1
26        1
44        1
50        1
60        1
165       1
84        1
88        1
144       1
182       1
143       1
666       1
960       1
1776      1
17        1
27        1
45        1
99        1
121       1
204       1
Name: rating_numerator, dtype: int64

In [359]:
ratings.rating_denominator.value_counts()

10     2333
11        3
50        3
80        2
20        2
2         1
16        1
40        1
70        1
15        1
90        1
110       1
120       1
130       1
150       1
170       1
7         1
0         1
Name: rating_denominator, dtype: int64

> Some rating values (numerator & denominator appears to be incorrect)

> Check for null values

In [360]:
df_twitter_archive.isnull().any()

tweet_id                      False
in_reply_to_status_id          True
in_reply_to_user_id            True
timestamp                     False
source                        False
text                          False
retweeted_status_id            True
retweeted_status_user_id       True
retweeted_status_timestamp     True
expanded_urls                  True
rating_numerator              False
rating_denominator            False
name                          False
doggo                         False
floofer                       False
pupper                        False
puppo                         False
dtype: bool

>

##### most of the columns containing null values are not unnecessary - there are basically id columns and retweets. 
##### The project instructions places emphasis on tweets and not retweets

> examine values of columns of dog typpes

In [361]:
df_twitter_archive[['doggo','floofer' ,'pupper', 'puppo']].values.tolist()[:3]

[['None', 'None', 'None', 'None'],
 ['None', 'None', 'None', 'None'],
 ['None', 'None', 'None', 'None']]

>  examine values in twweet source

In [362]:
df_twitter_archive.source.values

array(['<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>',
       '<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>',
       '<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>',
       ...,
       '<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>',
       '<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>',
       '<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>'],
      dtype=object)

In [363]:
df_twitter_archive.source.unique

<bound method Series.unique of 0       <a href="http://twitter.com/download/iphone" r...
1       <a href="http://twitter.com/download/iphone" r...
2       <a href="http://twitter.com/download/iphone" r...
3       <a href="http://twitter.com/download/iphone" r...
4       <a href="http://twitter.com/download/iphone" r...
                              ...                        
2351    <a href="http://twitter.com/download/iphone" r...
2352    <a href="http://twitter.com/download/iphone" r...
2353    <a href="http://twitter.com/download/iphone" r...
2354    <a href="http://twitter.com/download/iphone" r...
2355    <a href="http://twitter.com/download/iphone" r...
Name: source, Length: 2356, dtype: object>

In [364]:
# we only needed the client type (e.g iphone, androif, etc) of the tweet source not the link

In [365]:
# view exapnded urls column
df_twitter_archive.expanded_urls.values

array(['https://twitter.com/dog_rates/status/892420643555336193/photo/1',
       'https://twitter.com/dog_rates/status/892177421306343426/photo/1',
       'https://twitter.com/dog_rates/status/891815181378084864/photo/1',
       ...,
       'https://twitter.com/dog_rates/status/666033412701032449/photo/1',
       'https://twitter.com/dog_rates/status/666029285002620928/photo/1',
       'https://twitter.com/dog_rates/status/666020888022790149/photo/1'],
      dtype=object)

.

> Assessment 2: Image Predictions

In [366]:
df_img_pred

tweet_id                                          jpg_url  \
0     666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
...                  ...                                              ...   
2070  891327558926688256  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg   
2071  891689557279858688  https://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg   
2072  891815181378084864  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg   
2073  892177421306343426  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg   
2074  892420643555336193  https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg   

      img_num                      p1   p1_conf  p1_dog                  p2  \
0           1  Welsh_springer_spaniel  0.465074    True              collie   
1           1                 redbone  0.506826    True  miniature_pinscher   
2           1         German_shepherd  0.596461    True            malinois   
3           1     Rhodesian_ridgeback  0.408143    True             redbone   
4           1      miniature_pinscher  0.560311    True          Rottweiler   
...       ...                     ...       ...     ...                 ...   
2070        2                  basset  0.555712    True    English_springer   
2071        1             paper_towel  0.170278   False  Labrador_retriever   
2072        1               Chihuahua  0.716012    True            malamute   
2073        1               Chihuahua  0.323581    True            Pekinese   
2074        1                  orange  0.097049   False               bagel   

       p2_conf  p2_dog                           p3   p3_conf  p3_dog  
0     0.156665    True            Shetland_sheepdog  0.061428    True  
1     0.074192    True          Rhodesian_ridgeback  0.072010    True  
2     0.138584    True                   bloodhound  0.116197    True  
3     0.360687    True           miniature_pinscher  0.222752    True  
4     0.243682    True                     Doberman  0.154629    True  
...        ...     ...                          ...       ...     ...  
2070  0.225770    True  German_short-haired_pointer  0.175219    True  
2071  0.168086    True                      spatula  0.040836   False  
2072  0.078253    True                       kelpie  0.031379    True  
2073  0.090647    True                     papillon  0.068957    True  
2074  0.085851   False                       banana  0.076110   False  

[2075 rows x 12 columns]

> check unique size of dataset

In [367]:
df_img_pred.tweet_id.nunique()

2075

In [368]:
df_img_pred.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


> check for null values

In [369]:
df_img_pred.isnull().any().sum()

0

> describe values

In [370]:
df_img_pred.describe()

tweet_id      img_num      p1_conf       p2_conf       p3_conf
count  2.075000e+03  2075.000000  2075.000000  2.075000e+03  2.075000e+03
mean   7.384514e+17     1.203855     0.594548  1.345886e-01  6.032417e-02
std    6.785203e+16     0.561875     0.271174  1.006657e-01  5.090593e-02
min    6.660209e+17     1.000000     0.044333  1.011300e-08  1.740170e-10
25%    6.764835e+17     1.000000     0.364412  5.388625e-02  1.622240e-02
50%    7.119988e+17     1.000000     0.588230  1.181810e-01  4.944380e-02
75%    7.932034e+17     1.000000     0.843855  1.955655e-01  9.180755e-02
max    8.924206e+17     4.000000     1.000000  4.880140e-01  2.734190e-01

> examine the p1, p2 and p3 columns

In [371]:
df_img_pred[['p1', 'p2', 'p3']]

p1                  p2                           p3
0     Welsh_springer_spaniel              collie            Shetland_sheepdog
1                    redbone  miniature_pinscher          Rhodesian_ridgeback
2            German_shepherd            malinois                   bloodhound
3        Rhodesian_ridgeback             redbone           miniature_pinscher
4         miniature_pinscher          Rottweiler                     Doberman
...                      ...                 ...                          ...
2070                  basset    English_springer  German_short-haired_pointer
2071             paper_towel  Labrador_retriever                      spatula
2072               Chihuahua            malamute                       kelpie
2073               Chihuahua            Pekinese                     papillon
2074                  orange               bagel                       banana

[2075 rows x 3 columns]

In [372]:
df_img_pred[['p1', 'p2', 'p3']].describe()

p1                  p2                  p3
count               2075                2075                2075
unique               378                 405                 408
top     golden_retriever  Labrador_retriever  Labrador_retriever
freq                 150                 104                  79

.

> Assessment 3: Json Tweets

In [373]:
df_tweets

created_at                  id              id_str  \
0     Tue Aug 01 16:23:56 +0000 2017  892420643555336193  892420643555336193   
1     Tue Aug 01 00:17:27 +0000 2017  892177421306343426  892177421306343426   
2     Mon Jul 31 00:18:03 +0000 2017  891815181378084864  891815181378084864   
3     Sun Jul 30 15:58:51 +0000 2017  891689557279858688  891689557279858688   
4     Sat Jul 29 16:00:24 +0000 2017  891327558926688256  891327558926688256   
...                              ...                 ...                 ...   
2349  Mon Nov 16 00:24:50 +0000 2015  666049248165822465  666049248165822465   
2350  Mon Nov 16 00:04:52 +0000 2015  666044226329800704  666044226329800704   
2351  Sun Nov 15 23:21:54 +0000 2015  666033412701032449  666033412701032449   
2352  Sun Nov 15 23:05:30 +0000 2015  666029285002620928  666029285002620928   
2353  Sun Nov 15 22:32:08 +0000 2015  666020888022790149  666020888022790149   

                                              full_text  truncated  \
0     This is Phineas. He's a mystical boy. Only eve...      False   
1     This is Tilly. She's just checking pup on you....      False   
2     This is Archie. He is a rare Norwegian Pouncin...      False   
3     This is Darla. She commenced a snooze mid meal...      False   
4     This is Franklin. He would like you to stop ca...      False   
...                                                 ...        ...   
2349  Here we have a 1949 1st generation vulpix. Enj...      False   
2350  This is a purebred Piers Morgan. Loves to Netf...      False   
2351  Here is a very happy pup. Big fan of well-main...      False   
2352  This is a western brown Mitsubishi terrier. Up...      False   
2353  Here we have a Japanese Irish Setter. Lost eye...      False   

     display_text_range                                           entities  \
0               [0, 85]  {'hashtags': [], 'symbols': [], 'user_mentions...   
1              [0, 138]  {'hashtags': [], 'symbols': [], 'user_mentions...   
2              [0, 121]  {'hashtags': [], 'symbols': [], 'user_mentions...   
3               [0, 79]  {'hashtags': [], 'symbols': [], 'user_mentions...   
4              [0, 138]  {'hashtags': [{'text': 'BarkWeek', 'indices': ...   
...                 ...                                                ...   
2349           [0, 120]  {'hashtags': [], 'symbols': [], 'user_mentions...   
2350           [0, 137]  {'hashtags': [], 'symbols': [], 'user_mentions...   
2351           [0, 130]  {'hashtags': [], 'symbols': [], 'user_mentions...   
2352           [0, 139]  {'hashtags': [], 'symbols': [], 'user_mentions...   
2353           [0, 131]  {'hashtags': [], 'symbols': [], 'user_mentions...   

                                      extended_entities  \
0     {'media': [{'id': 892420639486877696, 'id_str'...   
1     {'media': [{'id': 892177413194625024, 'id_str'...   
2     {'media': [{'id': 891815175371796480, 'id_str'...   
3     {'media': [{'id': 891689552724799489, 'id_str'...   
4     {'media': [{'id': 891327551943041024, 'id_str'...   
...                                                 ...   
2349  {'media': [{'id': 666049244999131136, 'id_str'...   
2350  {'media': [{'id': 666044217047650304, 'id_str'...   
2351  {'media': [{'id': 666033409081393153, 'id_str'...   
2352  {'media': [{'id': 666029276303482880, 'id_str'...   
2353  {'media': [{'id': 666020881337073664, 'id_str'...   

                                                 source  \
0     <a href="http://twitter.com/download/iphone" r...   
1     <a href="http://twitter.com/download/iphone" r...   
2     <a href="http://twitter.com/download/iphone" r...   
3     <a href="http://twitter.com/download/iphone" r...   
4     <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
2349  <a href="http://twitter.com/download/iphone" r...   
2350  <a href="http://twitter.com/download/iphone" r...   
2351  <a href="http://twitter.com/download/ipho

In [374]:
# gather info about data
df_tweets.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2354 entries, 0 to 2353
Data columns (total 31 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   created_at                     2354 non-null   object 
 1   id                             2354 non-null   int64  
 2   id_str                         2354 non-null   object 
 3   full_text                      2354 non-null   object 
 4   truncated                      2354 non-null   bool   
 5   display_text_range             2354 non-null   object 
 6   entities                       2354 non-null   object 
 7   extended_entities              2073 non-null   object 
 8   source                         2354 non-null   object 
 9   in_reply_to_status_id          78 non-null     float64
 10  in_reply_to_status_id_str      78 non-null     object 
 11  in_reply_to_user_id            78 non-null     float64
 12  in_reply_to_user_id_str        78 non-null     o

In [375]:
# check data tyoes
df_tweets.dtypes

created_at                        object
id                                 int64
id_str                            object
full_text                         object
truncated                           bool
display_text_range                object
entities                          object
extended_entities                 object
source                            object
in_reply_to_status_id            float64
in_reply_to_status_id_str         object
in_reply_to_user_id              float64
in_reply_to_user_id_str           object
in_reply_to_screen_name           object
user                              object
geo                               object
coordinates                       object
place                             object
contributors                      object
is_quote_status                     bool
retweet_count                      int64
favorite_count                     int64
favorited                           bool
retweeted                           bool
possibly_sensiti

In [376]:
# find number of retweets
df_tweets.query('retweeted == False').shape[0]

2354

###### _All Tweets here are not retweets_

In [377]:
# examin language columns
#### we might have to create a dictionary to map the full name of the language to thier abbreviations
df_tweets.lang.value_counts()

en     2336
und       7
nl        3
in        3
tl        1
es        1
et        1
ro        1
eu        1
Name: lang, dtype: int64

In [378]:
# examine qntities columns
df_tweets[['entities','extended_entities']]

entities  \
0     {'hashtags': [], 'symbols': [], 'user_mentions...   
1     {'hashtags': [], 'symbols': [], 'user_mentions...   
2     {'hashtags': [], 'symbols': [], 'user_mentions...   
3     {'hashtags': [], 'symbols': [], 'user_mentions...   
4     {'hashtags': [{'text': 'BarkWeek', 'indices': ...   
...                                                 ...   
2349  {'hashtags': [], 'symbols': [], 'user_mentions...   
2350  {'hashtags': [], 'symbols': [], 'user_mentions...   
2351  {'hashtags': [], 'symbols': [], 'user_mentions...   
2352  {'hashtags': [], 'symbols': [], 'user_mentions...   
2353  {'hashtags': [], 'symbols': [], 'user_mentions...   

                                      extended_entities  
0     {'media': [{'id': 892420639486877696, 'id_str'...  
1     {'media': [{'id': 892177413194625024, 'id_str'...  
2     {'media': [{'id': 891815175371796480, 'id_str'...  
3     {'media': [{'id': 891689552724799489, 'id_str'...  
4     {'media': [{'id': 891327551943041024, 'id_str'...  
...                                                 ...  
2349  {'media': [{'id': 666049244999131136, 'id_str'...  
2350  {'media': [{'id': 666044217047650304, 'id_str'...  
2351  {'media': [{'id': 666033409081393153, 'id_str'...  
2352  {'media': [{'id': 666029276303482880, 'id_str'...  
2353  {'media': [{'id': 666020881337073664, 'id_str'...  

[2354 rows x 2 columns]

> at minimum the tweets dataframe should contain the following columns,

>  tweet ID, retweet count, and favorite count
> hence we would examine them

In [379]:
main_tweet_cols = df_tweets[['id','retweet_count','favorite_count']]
main_tweet_cols.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2354 entries, 0 to 2353
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype
---  ------          --------------  -----
 0   id              2354 non-null   int64
 1   retweet_count   2354 non-null   int64
 2   favorite_count  2354 non-null   int64
dtypes: int64(3)
memory usage: 55.3 KB


> since these columns look preety good (do not have null values), 

> we might as well use only them and a few other columns that are also complete

#### Assessment Summary

#### Quality Issues

##### Twitter Archives
  > 1. Some columns (_retweets & ids_) contain null values and are not really needed
  > 2. Columns on dog stages (_doggy, pupper etc_) contain mostly NUll values
  > 3. source column values not in correct format
  > 4. expanded urls column not needed. The only useful information needed from it (tweet id) is already availabele
       in a separate column
  > 5. all retweets columns (_retweeted_status_id,retweeted_status_user_id,retweeted_status_timestamp_) not
      required
  > 6  Incorrect denominator rating values, values greater than 10 and some quite outrageous
  > 7  Incorrect numerator rating values; we have outrageous values such as 420, 1776
##### Image Predictions
  > 8. The string values in p1, p2, and p3 coluumns _breed predictions_ by the neural network are not in uniform format
  > 9   Unique counts of ids is less than that in archives - missing data
  > 10   columns such as tweet_id are not in thier corect datatype
##### Json Tweets
  > 11. The column id datatype in int instead of string
  > 12. language column values are not meaningful
  

#### Tidiness
##### Twitter Archives
 > 1. dog stages in different columns in twitter_archve dta2etimage prediction datasets 
 ##### Image Predictions
 > 2. image prediction datasets needs to be joined with twitter archive rather than in separate datasets
   genereally we have 3 separe column datasets rather than one master dataset
 ##### Json Tweets
 > 3. multiple id columns in json_tweets data
 > 4. columns like (_truncated 	display_text_range 	entities 	extended_entities 	source 	in_reply_to_status_id_)
       not required in terms of analysis
 > 5. datasets needs to be joined with twitter archive data

.

### Cleaning Data

##### Clean each of the issues you documented while assessing. Perform this cleaning in wrangle_act.ipynb as well. 
##### The result should be a high quality and tidy master pandas DataFrame (or DataFrames, if appropriate). 
##### Again, the issues that satisfy the Project Motivation must be cleaned.

> Create copies of Dataframe before cleaning

In [380]:
twitter_archive_copy = df_twitter_archive.copy()

In [381]:
img_pred_copy = df_img_pred.copy()

In [382]:
tweets_copy = df_tweets.copy()

> Load datasets copy

In [383]:
twitter_archive_copy.head(1)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   

   rating_denominator     name doggo floofer pupper puppo  
0                  10  Phineas  None    None   None  None

.

In [384]:
img_pred_copy.head(1)

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   

   img_num                      p1   p1_conf  p1_dog      p2   p2_conf  \
0        1  Welsh_springer_spaniel  0.465074    True  collie  0.156665   

   p2_dog                 p3   p3_conf  p3_dog  
0    True  Shetland_sheepdog  0.061428    True

.

In [385]:
tweets_copy.head(1)

created_at                  id              id_str  \
0  Tue Aug 01 16:23:56 +0000 2017  892420643555336193  892420643555336193   

                                           full_text  truncated  \
0  This is Phineas. He's a mystical boy. Only eve...      False   

  display_text_range                                           entities  \
0            [0, 85]  {'hashtags': [], 'symbols': [], 'user_mentions...   

                                   extended_entities  \
0  {'media': [{'id': 892420639486877696, 'id_str'...   

                                              source  in_reply_to_status_id  \
0  <a href="http://twitter.com/download/iphone" r...                    NaN   

   ... favorite_count  favorited retweeted possibly_sensitive  \
0  ...          39467      False     False              False   

  possibly_sensitive_appealable lang retweeted_status quoted_status_id  \
0                         False   en              NaN              NaN   

  quoted_status_id_str  quoted_status  
0                  NaN            NaN  

[1 rows x 31 columns]

##### Clean Twitter Archives
         Quality issues

  > 1. Some columns (_retweets & ids_) contain null values and are not really needed
  > 2. Columns on dog stages (_doggy, pupper etc_) contain mostly NUll values
  > 3. source column values not in correct format
  > 4. expanded urls column not needed. The only useful information needed from it (tweet id) is already availabele
       in a separate column
  > 5. all retweets columns (_retweeted_status_id,retweeted_status_user_id,retweeted_status_timestamp_) not
      required
  > 6  Incorrect denominator rating values, values greater than 10 and some quite outrageous
  > 7  Incorrect numerator rating values; we have outrageous values such as 420, 1776

columns with null values 
> Define

>>Drop columns with null values which are also not useful for analysis



   

> Code

In [386]:
twitter_archive_copy.isnull().any()

tweet_id                      False
in_reply_to_status_id          True
in_reply_to_user_id            True
timestamp                     False
source                        False
text                          False
retweeted_status_id            True
retweeted_status_user_id       True
retweeted_status_timestamp     True
expanded_urls                  True
rating_numerator              False
rating_denominator            False
name                          False
doggo                         False
floofer                       False
pupper                        False
puppo                         False
dtype: bool

In [387]:
twitter_archive_copy.columns

Index(['tweet_id', 'in_reply_to_status_id', 'in_reply_to_user_id', 'timestamp',
       'source', 'text', 'retweeted_status_id', 'retweeted_status_user_id',
       'retweeted_status_timestamp', 'expanded_urls', 'rating_numerator',
       'rating_denominator', 'name', 'doggo', 'floofer', 'pupper', 'puppo'],
      dtype='object')

In [388]:
drop_cols = ['in_reply_to_status_id', 'in_reply_to_user_id','retweeted_status_id', 'retweeted_status_user_id',
       'retweeted_status_timestamp', 'expanded_urls']
twitter_archive_copy.drop(columns=drop_cols, inplace=True)

> Test 

In [389]:
twitter_archive_copy.isnull().any()

tweet_id              False
timestamp             False
source                False
text                  False
rating_numerator      False
rating_denominator    False
name                  False
doggo                 False
floofer               False
pupper                False
puppo                 False
dtype: bool

In [390]:
twitter_archive_copy

tweet_id                  timestamp  \
0     892420643555336193  2017-08-01 16:23:56 +0000   
1     892177421306343426  2017-08-01 00:17:27 +0000   
2     891815181378084864  2017-07-31 00:18:03 +0000   
3     891689557279858688  2017-07-30 15:58:51 +0000   
4     891327558926688256  2017-07-29 16:00:24 +0000   
...                  ...                        ...   
2351  666049248165822465  2015-11-16 00:24:50 +0000   
2352  666044226329800704  2015-11-16 00:04:52 +0000   
2353  666033412701032449  2015-11-15 23:21:54 +0000   
2354  666029285002620928  2015-11-15 23:05:30 +0000   
2355  666020888022790149  2015-11-15 22:32:08 +0000   

                                                 source  \
0     <a href="http://twitter.com/download/iphone" r...   
1     <a href="http://twitter.com/download/iphone" r...   
2     <a href="http://twitter.com/download/iphone" r...   
3     <a href="http://twitter.com/download/iphone" r...   
4     <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
2351  <a href="http://twitter.com/download/iphone" r...   
2352  <a href="http://twitter.com/download/iphone" r...   
2353  <a href="http://twitter.com/download/iphone" r...   
2354  <a href="http://twitter.com/download/iphone" r...   
2355  <a href="http://twitter.com/download/iphone" r...   

                                                   text  rating_numerator  \
0     This is Phineas. He's a mystical boy. Only eve...                13   
1     This is Tilly. She's just checking pup on you....                13   
2     This is Archie. He is a rare Norwegian Pouncin...                12   
3     This is Darla. She commenced a snooze mid meal...                13   
4     This is Franklin. He would like you to stop ca...                12   
...                                                 ...               ...   
2351  Here we have a 1949 1st generation vulpix. Enj...                 5   
2352  This is a purebred Piers Morgan. Loves to Netf...                 6   
2353  Here is a very happy pup. Big fan of well-main...                 9   
2354  This is a western brown Mitsubishi terrier. Up...                 7   
2355  Here we have a Japanese Irish Setter. Lost eye...                 8   

      rating_denominator      name doggo floofer pupper puppo  
0                     10   Phineas  None    None   None  None  
1                     10     Tilly  None    None   None  None  
2                     10    Archie  None    None   None  None  
3                     10     Darla  None    None   None  None  
4                     10  Franklin  None    None   None  None  
...                  ...       ...   ...     ...    ...   ...  
2351                  10      None  None    None   None  None  
2352                  10         a  None    None   None  None  
2353                  10         a  None    None   None  None  
2354                  10         a  None    None   None  None  
2355                  10      None  None    None   None  None  

[2356 rows x 11 columns]

Dog Stages Columns with None Values


> Define 

Inspect text column to check if dog stages name wwas or was not properly extracted

If extraction is feasible, use the extract function and regex to get the dog stage type out

In [391]:
twitter_archive_copy.doggo.apply(lambda x: "doggo " in twitter_archive_copy.text and  "doggo" or  "None").value_counts()

None    2356
Name: doggo, dtype: int64

Source column values not in correct format
> Define

extract tweet source (i-phone, etc) from url string using pandas extract function

> Code

In [500]:
twitter_archive_copy.source.str.split(pat="for ", expand=True)[1]\
.str.split(pat="<", expand=True)[0].unique()

array(['iPhone', None], dtype=object)

In [436]:
twitter_archive_copy['rsource'] = twitter_archive_copy.source.str.split('for ').tolist

In [502]:
twitter_archive_copy.source.unique()

array(['<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>',
       '<a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>',
       '<a href="http://vine.co" rel="nofollow">Vine - Make a Scene</a>',
       '<a href="https://about.twitter.com/products/tweetdeck" rel="nofollow">TweetDeck</a>'],
      dtype=object)

In [522]:
import re
sr = re.compile('>(iPhone|WebClient|TweetDeck)')

In [523]:
twitter_archive_copy.source.apply(lambda x: sr.match(x)).value_counts()

Series([], Name: source, dtype: int64)

In [438]:
twitter_archive_copy.rsource.values

array([list(['<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter ', 'iPhone</a>']),
       list(['<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter ', 'iPhone</a>']),
       list(['<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter ', 'iPhone</a>']),
       ...,
       list(['<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter ', 'iPhone</a>']),
       list(['<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter ', 'iPhone</a>']),
       list(['<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter ', 'iPhone</a>'])],
      dtype=object)

###  Store, Analyze & Visualize Data

#### Store the clean DataFrame(s) in a CSV file with the main one named `twitter_archive_master.csv`. 
#### If additional files exist because multiple tables are required for tidiness, name these files appropriately. 
#### Additionally, you may store the cleaned data in a SQLite database (which is to be submitted as well if you do).
#### Analyze and visualize your wrangled data in your wrangle_act.ipynb Jupyter Notebook. 
#### `At least three (3) insights and one (1) visualization must be produced`.